File: packages/shared-sdk/src/cache/redis-cache.ts
/**
 * Redis Cache Implementation for Aviation SDK
 *
 * Provides a unified caching interface with TTL management,
 * hit rate monitoring, and automatic serialization.
 *
 * @module @aviation/shared-sdk/cache
 */

import { createClient, RedisClientType } from 'redis';

export interface CacheConfig {
  host?: string;
  port?: number;
  password?: string;
  db?: number;
  keyPrefix?: string;
  defaultTTL?: number;
  enableMetrics?: boolean;
}

export interface CacheMetrics {
  hits: number;
  misses: number;
  sets: number;
  deletes: number;
  hitRate: number;
}

export interface CacheOptions {
  ttl?: number; // Seconds
  serialize?: boolean;
}

/**
 * Redis-based cache with metrics and TTL management
 */
export class RedisCache {
  private client: RedisClientType;
  private config: Required<CacheConfig>;
  private metrics: CacheMetrics = {
    hits: 0,
    misses: 0,
    sets: 0,
    deletes: 0,
    hitRate: 0,
  };
  private connected: boolean = false;

  constructor(config: CacheConfig = {}) {
    this.config = {
      host: config.host || process.env.REDIS_HOST || 'localhost',
      port: config.port || parseInt(process.env.REDIS_PORT || '6379'),
      password: config.password || process.env.REDIS_PASSWORD || '',
      db: config.db || 0,
      keyPrefix: config.keyPrefix || 'aviation:',
      defaultTTL: config.defaultTTL || 3600, // 1 hour default
      enableMetrics: config.enableMetrics !== false,
    };

    this.client = createClient({
      socket: {
        host: this.config.host,
        port: this.config.port,
      },
      password: this.config.password || undefined,
      database: this.config.db,
    });

    this.client.on('error', (err) => console.error('Redis Client Error:', err));
    this.client.on('connect', () => {
      this.connected = true;
      console.log('✅ Redis cache connected');
    });
  }

  /**
   * Connect to Redis
   */
  async connect(): Promise<void> {
    if (!this.connected) {
      await this.client.connect();
    }
  }

  /**
   * Disconnect from Redis
   */
  async disconnect(): Promise<void> {
    if (this.connected) {
      await this.client.quit();
      this.connected = false;
    }
  }

  /**
   * Generate full cache key with prefix
   */
  private key(key: string): string {
    return `${this.config.keyPrefix}${key}`;
  }

  /**
   * Get value from cache
   */
  async get<T>(key: string, options: CacheOptions = {}): Promise<T | null> {
    try {
      const value = await this.client.get(this.key(key));

      if (value === null) {
        if (this.config.enableMetrics) this.metrics.misses++;
        return null;
      }

      if (this.config.enableMetrics) this.metrics.hits++;

      return options.serialize !== false ? JSON.parse(value) : (value as unknown as T);
    } catch (error) {
      console.error('Cache get error:', error);
      return null;
    }
  }

  /**
   * Set value in cache with TTL
   */
  async set<T>(key: string, value: T, options: CacheOptions = {}): Promise<boolean> {
    try {
      const ttl = options.ttl || this.config.defaultTTL;
      const serialized = options.serialize !== false ? JSON.stringify(value) : String(value);

      await this.client.setEx(this.key(key), ttl, serialized);

      if (this.config.enableMetrics) this.metrics.sets++;
      return true;
    } catch (error) {
      console.error('Cache set error:', error);
      return false;
    }
  }

  /**
   * Get or set pattern - execute function if cache miss
   */
  async getOrSet<T>(
    key: string,
    fn: () => Promise<T>,
    options: CacheOptions = {}
  ): Promise<T> {
    const cached = await this.get<T>(key, options);

    if (cached !== null) {
      return cached;
    }

    const value = await fn();
    await this.set(key, value, options);

    return value;
  }

  /**
   * Delete key from cache
   */
  async delete(key: string): Promise<boolean> {
    try {
      const result = await this.client.del(this.key(key));
      if (this.config.enableMetrics) this.metrics.deletes++;
      return result > 0;
    } catch (error) {
      console.error('Cache delete error:', error);
      return false;
    }
  }

  /**
   * Delete keys matching pattern
   */
  async deletePattern(pattern: string): Promise<number> {
    try {
      const keys = await this.client.keys(this.key(pattern));
      if (keys.length === 0) return 0;

      const result = await this.client.del(keys);
      if (this.config.enableMetrics) this.metrics.deletes += result;
      return result;
    } catch (error) {
      console.error('Cache delete pattern error:', error);
      return 0;
    }
  }

  /**
   * Check if key exists
   */
  async exists(key: string): Promise<boolean> {
    try {
      const result = await this.client.exists(this.key(key));
      return result > 0;
    } catch (error) {
      console.error('Cache exists error:', error);
      return false;
    }
  }

  /**
   * Get remaining TTL for key (in seconds)
   */
  async ttl(key: string): Promise<number> {
    try {
      return await this.client.ttl(this.key(key));
    } catch (error) {
      console.error('Cache TTL error:', error);
      return -1;
    }
  }

  /**
   * Extend TTL for existing key
   */
  async extend(key: string, seconds: number): Promise<boolean> {
    try {
      const result = await this.client.expire(this.key(key), seconds);
      return result;
    } catch (error) {
      console.error('Cache extend error:', error);
      return false;
    }
  }

  /**
   * Get cache metrics
   */
  getMetrics(): CacheMetrics {
    const total = this.metrics.hits + this.metrics.misses;
    this.metrics.hitRate = total > 0 ? (this.metrics.hits / total) * 100 : 0;
    return { ...this.metrics };
  }

  /**
   * Reset cache metrics
   */
  resetMetrics(): void {
    this.metrics = {
      hits: 0,
      misses: 0,
      sets: 0,
      deletes: 0,
      hitRate: 0,
    };
  }

  /**
   * Flush entire cache (use with caution!)
   */
  async flush(): Promise<void> {
    try {
      await this.client.flushDb();
      console.warn('⚠️  Cache flushed');
    } catch (error) {
      console.error('Cache flush error:', error);
    }
  }

  /**
   * Get cache info
   */
  async info(): Promise<string> {
    try {
      return await this.client.info();
    } catch (error) {
      console.error('Cache info error:', error);
      return '';
    }
  }
}

/**
 * Cache TTL constants for different data types
 */
export const CacheTTL = {
  AIRPORT_DATA: 86400, // 24 hours (rarely changes)
  WEATHER_CURRENT: 900, // 15 minutes
  WEATHER_FORECAST: 1800, // 30 minutes
  METAR: 1800, // 30 minutes
  TAF: 3600, // 1 hour
  ROUTE_CALCULATION: 3600, // 1 hour
  MAP_TILES: 604800, // 7 days
  STATIC_ASSETS: 2592000, // 30 days
  USER_SESSION: 86400, // 24 hours
  RATE_LIMIT: 60, // 1 minute
} as const;

/**
 * Global cache instance (singleton)
 */
let globalCache: RedisCache | null = null;

/**
 * Get or create global cache instance
 */
export function getCache(config?: CacheConfig): RedisCache {
  if (!globalCache) {
    globalCache = new RedisCache(config);
  }
  return globalCache;
}

/**
 * Initialize and connect cache
 */
export async function initCache(config?: CacheConfig): Promise<RedisCache> {
  const cache = getCache(config);
  await cache.connect();
  return cache;
}

File: packages/shared-sdk/python/aviation/cache/redis_cache.py
"""
Redis Cache Implementation for Aviation Python SDK

Provides a unified caching interface with TTL management,
hit rate monitoring, and automatic serialization.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from typing import Any, Callable, Optional, TypeVar

try:
    import redis
    from redis import Redis
except ImportError:
    redis = None  # type: ignore
    Redis = None  # type: ignore

T = TypeVar("T")


@dataclass
class CacheConfig:
    """Redis cache configuration"""

    host: str = field(default_factory=lambda: os.getenv("REDIS_HOST", "localhost"))
    port: int = field(default_factory=lambda: int(os.getenv("REDIS_PORT", "6379")))
    password: Optional[str] = field(default_factory=lambda: os.getenv("REDIS_PASSWORD"))
    db: int = 0
    key_prefix: str = "aviation:"
    default_ttl: int = 3600  # 1 hour
    enable_metrics: bool = True


@dataclass
class CacheMetrics:
    """Cache performance metrics"""

    hits: int = 0
    misses: int = 0
    sets: int = 0
    deletes: int = 0

    @property
    def hit_rate(self) -> float:
        """Calculate cache hit rate percentage"""
        total = self.hits + self.misses
        return (self.hits / total * 100) if total > 0 else 0.0


class RedisCache:
    """Redis-based cache with metrics and TTL management"""

    def __init__(self, config: Optional[CacheConfig] = None):
        if redis is None:
            raise RuntimeError(
                "redis package not installed. Install with: pip install redis"
            )

        self.config = config or CacheConfig()
        self.metrics = CacheMetrics()
        self._client: Optional[Redis] = None

    def connect(self) -> None:
        """Connect to Redis"""
        if self._client is None:
            self._client = redis.Redis(
                host=self.config.host,
                port=self.config.port,
                password=self.config.password,
                db=self.config.db,
                decode_responses=True,
            )
            # Test connection
            self._client.ping()
            print("✅ Redis cache connected")

    def disconnect(self) -> None:
        """Disconnect from Redis"""
        if self._client:
            self._client.close()
            self._client = None

    @property
    def client(self) -> Redis:
        """Get Redis client, connecting if necessary"""
        if self._client is None:
            self.connect()
        return self._client  # type: ignore

    def _key(self, key: str) -> str:
        """Generate full cache key with prefix"""
        return f"{self.config.key_prefix}{key}"

    def get(self, key: str, serialize: bool = True) -> Optional[Any]:
        """Get value from cache"""
        try:
            value = self.client.get(self._key(key))

            if value is None:
                if self.config.enable_metrics:
                    self.metrics.misses += 1
                return None

            if self.config.enable_metrics:
                self.metrics.hits += 1

            return json.loads(value) if serialize else value
        except Exception as e:
            print(f"Cache get error: {e}")
            return None

    def set(
        self, key: str, value: Any, ttl: Optional[int] = None, serialize: bool = True
    ) -> bool:
        """Set value in cache with TTL"""
        try:
            ttl = ttl if ttl is not None else self.config.default_ttl
            serialized = json.dumps(value) if serialize else str(value)

            self.client.setex(self._key(key), ttl, serialized)

            if self.config.enable_metrics:
                self.metrics.sets += 1
            return True
        except Exception as e:
            print(f"Cache set error: {e}")
            return False

    def get_or_set(
        self, key: str, fn: Callable[[], T], ttl: Optional[int] = None, serialize: bool = True
    ) -> T:
        """Get or set pattern - execute function if cache miss"""
        cached = self.get(key, serialize=serialize)

        if cached is not None:
            return cached

        value = fn()
        self.set(key, value, ttl=ttl, serialize=serialize)

        return value

    def delete(self, key: str) -> bool:
        """Delete key from cache"""
        try:
            result = self.client.delete(self._key(key))
            if self.config.enable_metrics:
                self.metrics.deletes += 1
            return result > 0
        except Exception as e:
            print(f"Cache delete error: {e}")
            return False

    def delete_pattern(self, pattern: str) -> int:
        """Delete keys matching pattern"""
        try:
            keys = self.client.keys(self._key(pattern))
            if not keys:
                return 0

            result = self.client.delete(*keys)
            if self.config.enable_metrics:
                self.metrics.deletes += result
            return result
        except Exception as e:
            print(f"Cache delete pattern error: {e}")
            return 0

    def exists(self, key: str) -> bool:
        """Check if key exists"""
        try:
            return self.client.exists(self._key(key)) > 0
        except Exception as e:
            print(f"Cache exists error: {e}")
            return False

    def ttl(self, key: str) -> int:
        """Get remaining TTL for key (in seconds)"""
        try:
            return self.client.ttl(self._key(key))
        except Exception as e:
            print(f"Cache TTL error: {e}")
            return -1

    def extend(self, key: str, seconds: int) -> bool:
        """Extend TTL for existing key"""
        try:
            return self.client.expire(self._key(key), seconds)
        except Exception as e:
            print(f"Cache extend error: {e}")
            return False

    def get_metrics(self) -> CacheMetrics:
        """Get cache metrics"""
        return self.metrics

    def reset_metrics(self) -> None:
        """Reset cache metrics"""
        self.metrics = CacheMetrics()

    def flush(self) -> None:
        """Flush entire cache (use with caution!)"""
        try:
            self.client.flushdb()
            print("⚠️  Cache flushed")
        except Exception as e:
            print(f"Cache flush error: {e}")

    def info(self) -> str:
        """Get cache info"""
        try:
            return str(self.client.info())
        except Exception as e:
            print(f"Cache info error: {e}")
            return ""


# Cache TTL constants (in seconds)
class CacheTTL:
    """Standard TTL values for different data types"""

    AIRPORT_DATA = 86400  # 24 hours (rarely changes)
    WEATHER_CURRENT = 900  # 15 minutes
    WEATHER_FORECAST = 1800  # 30 minutes
    METAR = 1800  # 30 minutes
    TAF = 3600  # 1 hour
    ROUTE_CALCULATION = 3600  # 1 hour
    MAP_TILES = 604800  # 7 days
    STATIC_ASSETS = 2592000  # 30 days
    USER_SESSION = 86400  # 24 hours
    RATE_LIMIT = 60  # 1 minute


# Global cache instance (singleton)
_global_cache: Optional[RedisCache] = None


def get_cache(config: Optional[CacheConfig] = None) -> RedisCache:
    """Get or create global cache instance"""
    global _global_cache
    if _global_cache is None:
        _global_cache = RedisCache(config)
    return _global_cache


def init_cache(config: Optional[CacheConfig] = None) -> RedisCache:
    """Initialize and connect cache"""
    cache = get_cache(config)
    cache.connect()
    return cache
